DMA solution to glitching/tearing problem

theredwagoneer opened the pull request:
This solution is an alternate to https://github.com/blaz-r/pi_pico_neopixel/pull/24.  It also solves the glitching problem, but without disabling interrupts.

theredwagoneer committed:
DMA solution to glitching/tearing problem

blaz-r commented:
Hi, thanks for this effort 😄 
This code does work fine for me, but only if the leds are on pin 0. Could you look into this please? 
Also I think DMA option should be optional, so could you also make it a flag that you pass to the class? Thank you,

theredwagoneer committed:
Merge branch 'main' into dma_solution

theredwagoneer committed:
Added options for transfer mode

theredwagoneer committed:
Fixed so it works on alternate pins

theredwagoneer commented:
Yes, I see the problem.  I have the fix and the options implemented, but it may be a few days before I can test it on the HW.

blaz-r commented on neopixel.py:
```diff
@@ -342,7 +345,18 @@ def show(self):
             cut = 0
         
-        self.sm.put(self.pixels, cut)
+        data = array.array('I',self.pixels)
+        for i,_ in enumerate(data):
+            data[i] <<= cut
```
I think this should also be merged with the option above where cut is 0, so we avoid unnecessary copy in case of RGBW. Or is this necessary?

blaz-r commented on neopixel.py:
```diff
@@ -101,6 +101,9 @@ def __init__(self, num_leds, state_machine, pin, mode="RGB", delay=0.0003):
         self.num_leds = num_leds
         self.delay = delay
         self.brightnessvalue = 255
+        self.dma = rp2.DMA()
+        DATA_REQUEST_INDEX = (pin << 3) + state_machine
+        self.dma_ctrl = self.dma.pack_ctrl(size=2, inc_write=False, treq_sel=DATA_REQUEST_INDEX)
```
It would be also nice to add some comment here for later development. Right now, it's not entirely clear from the code alone where these indexes come from.

blaz-r commented:
> Yes, I see the problem. I have the fix and the options implemented, but it may be a few days before I can test it on the HW.

Thanks. No worries, whenever you get some time. I now have access to my pico at all times so I can verify it quicker.

theredwagoneer committed:
Fix DREQ calculation
Add comment about how DREQ is calculated
Add options for disable interrupt or DMA

theredwagoneer commented:
Ok, I fixed the DREQ calculation and commented it.  I added an option to DMA or PUT or PUT _CRITICAL.  I also put the wait for the DMA in front instead of behind so that the user can take advantage of the transfer time if they want.

I test it on my Pico 2W on pin 0 with different PIOs.  I can't try other pins because my project is soldered together.  Can you try that?

blaz-r commented:
Great, thanks a lot. I'll check it out later today and I'll get back to you.

blaz-r committed:
Reintroduce data copy for DMA

blaz-r reviewed (approved):
Amazing work! I pushed a commit where I reintroduced copy of data even for RGBW, because it turned out that the pixels array could be changed while DMA reads it, leading to errors. Other than that, it now also works fine with pins other than 0 and the comment is well written.
Thank you for this contribution 😄

blaz-r commented:
Test if everything still works fine on your end and if it does, I'll merge.

theredwagoneer commented:
This looks good to me.  This is my first contribution to an open source project, so thanks for your support!

blaz-r commented:
Cool, then I'll merge. 🥳 Thanks again for your work.
Everyone starts somewhere, and for me this led library was one of the first real contributions as well 😄
